### PR TITLE
feat: BIO法

### PR DESCRIPTION
**Pull request by maton369**

### CODE REVIEW

**Review by maton369 — COMMENTED** [review]
GitHub Review — 8-23（BIO NER の学習・推論・評価：理論中心レビュー）

対象の要旨
- BIO ラベル空間（O=0, B-k=k, I-k=k+N, N=8 → 総ラベル=17）でトークン分類を学習（PyTorch Lightning）。
- ベスト ckpt をロードし、encode_plus_untagged のスパンとモデル出力ロジットから Viterbi 復号でエンティティ抽出。
- evaluate_model でスパン完全一致＋タイプ一致の micro Precision/Recall/F1 を算出。

────────────────────────────────────────────────────────────────────

■ 理論の妥当性：BIO・損失・系列整合

- ラベル定義は妥当である。num_labels = 2N+1 は BIO の標準的な符号化である。
- 学習損失は「各位置 t の多クラス交差エントロピーを時系列方向に平均」が定石であり、PAD/CLS/SEP は ignore_index=-100 により損失から除外する設計が合理的である。

$$
\mathcal{L}
= \frac{1}{|\mathcal{T}'|}
\sum_{t \in \mathcal{T}'}
-\log \frac{\exp\bigl(z_{t,\,y_t}\bigr)}
{\sum_{c=0}^{C-1}\exp\bigl(z_{t,\,c}\bigr)}\,,
$$

ここで \mathcal{T}' は「損失に含める位置集合」（PAD/CLS/SEP を除く）である。
- 推論では BIO の系列整合を後処理（Viterbi）で担保しており妥当である。ペナルティ行列で B/I 遷移制約を実装し、I の不正開始やタイプ跨ぎを大きな負ペナルティで抑制している。

────────────────────────────────────────────────────────────────────

■ 実装の良い点

- ラベル空間とモデル出力の整合性が明示されており、ヘッド出力次元と BIO 定義の齟齬が起きにくい。
- Viterbi 復号は実装が簡潔で、系列一貫性の確保に有効である。
- 評価設計（スパン完全一致＋タイプ一致の micro 集計）は研究・実務で広く用いられる基準である。

────────────────────────────────────────────────────────────────────

■ 主要なリスク・改善ポイント（理論と実装の接点）

1) Lightning v2 の API 変更  
- 現状の pl.Trainer(gpus=1, ...) は v2 系で非推奨/エラーとなる。可搬性（CUDA/MPS/CPU）確保のため accelerator='auto', devices=1 へ移行すべきである。

2) .cuda() 固定の可搬性  
- Mac(MPS)・CPU 環境で失敗する。モデル実デバイスに合わせて .to(device) に統一すべきである。

3) Viterbi のペナルティ・スケール  
- ペナルティはロジットのスケールに依存する。安定化のため、log-softmax（対数確率）へ正規化してから Viterbi に渡すことを推奨する。これによりペナルティを対数尤度差として解釈可能となる。

4) BIO 違反時の安全対策  
- Viterbi は違反を強く抑制するが絶対に禁止ではない。convert_bert_output_to_entities の I ブランチは「直前に entity が存在する」前提で更新しており、先頭 I で未定義参照の危険がある。先頭 I は B とみなして新規開始する等のガードを追加すべきである。

5) 前処理の一貫性  
- 学習で NFKC 正規化を適用しているため、推論時も同じ正規化を適用しないとスパンがずれ、完全一致の評価が過小になる。学習・推論で正規化を統一すべきである。

6) evaluate_model のゼロ割保護  
- 予測 0 件・正解 0 件時にゼロ割が発生し得る。分母が 0 の場合は 0.0 を返す等の保護が必須である。

7) 再現性・訓練安定化  
- pl.seed_everything(...)、AdamW + weight_decay、linear warmup/decay、gradient_clip_val、accumulate_grad_batches の導入により学習の安定性と再現性が向上する。

────────────────────────────────────────────────────────────────────

■ 最小パッチ（安全・可搬・理論整合の向上）

変更点（要点のみ。擬似コード体裁）：
- 変更前: trainer = pl.Trainer(gpus=1, max_epochs=5, callbacks=[checkpoint])
- 変更後: trainer = pl.Trainer(accelerator='auto', devices=1, max_epochs=5, callbacks=[checkpoint])

- 変更前: bert_tc = model.bert_tc.cuda()
- 変更後: 
  1) device を自動選択（MPS → CUDA → CPU）
  2) bert_tc = model.bert_tc.to(device).eval()

- 変更前: 推論で scores = output.logits[0].cpu().numpy().tolist() を直接 Viterbi に渡す
- 変更後: logits = output.logits[0] → scores = torch.log_softmax(logits, dim=-1).cpu().numpy().tolist() を Viterbi に渡す

- 変更前: encoding = {k: v.cuda() for k, v in encoding.items()}
- 変更後: encoding = {k: v.to(device) for k, v in encoding.items()}

- 変更前: print(evaluate_model(entities_list, entities_predicted_list))
- 変更後: ゼロ割保護済み evaluate_model を利用して print(...)

────────────────────────────────────────────────────────────────────

■ 評価設計の拡張（推奨）

- エンティティ F1（span-level）に加え、BIO ラベルのトークン F1 も併記すると境界誤差の傾向が把握しやすい（seqeval と torchmetrics の併用が便利である）。
- タイプ別（per-class）P/R/F1、macro/weighted-macro を併記し、不均衡データでの改善余地を可視化すべきである。
- 信頼度の較正（温度スケーリング）やタイプ別しきい値最適化で運用の安定性が向上する。

────────────────────────────────────────────────────────────────────

■ 最終コメント

- 現行実装は BIO の理論（系列制約）と BERT のトークン分類を正しく接続できている。  
- 可搬化（accelerator/devices, .to(device)）、Viterbi 前の log-softmax、I 先行のガード、評価時の正規化一貫性、ゼロ割保護を入れることで、安定性・再現性・評価の信頼性が一段上がる。  
- さらに AdamW＋ウォームアップや F1 ロギングを加えれば、学習曲線の解釈と改善サイクルがスムーズになる。
